### PR TITLE
1042 Run Pyright on diffs in CI

### LINKED ISSUE #1042: Check diffs with Pyright
Using https://github.com/Bachmann1234/diff_cover
And this plugin https://github.com/dperl-dls/pyright_diff_quality_plugin
We can test branches for Pyright type-checking compliance without having to fix everything in the codebase at once.

To test this out:
- Install both tools into a venv with hyperion
- switch to some branch which you think or know has errors
- run `diff-quality --violations pyright`

For example, on my current branch I get:

```
-------------
Diff Quality
Quality Report: pyright
Diff: origin/main...HEAD, staged and unstaged changes
-------------
conftest.py (100%)
src/__init__.py (100%)
src/hyperion/__main__.py (100%)
src/hyperion/external_interaction/callbacks/__init__.py (100%)
src/hyperion/external_interaction/callbacks/__main__.py (100%)
src/hyperion/external_interaction/callbacks/ispyb_callback_base.py (100%)
src/hyperion/external_interaction/callbacks/plan_reactive_callback.py (100%)
src/hyperion/external_interaction/callbacks/rotation/zocalo_callback.py (100%)
src/hyperion/external_interaction/callbacks/xray_centre/zocalo_callback.py (100%)
src/hyperion/external_interaction/zocalo/zocalo_interaction.py (85.7%):
src/hyperion/external_interaction/zocalo/zocalo_interaction.py:120: Object of type "None" is not subscriptable
src/hyperion/log.py (100%)
src/hyperion/parameters/cli.py (100%)
src/hyperion/parameters/constants.py (100%)
tests/system_tests/hyperion/test_main_system.py (100%)
tests/unit_tests/experiment_plans/conftest.py (100%)
tests/unit_tests/experiment_plans/test_flyscan_xray_centre_plan.py (100%)
tests/unit_tests/external_interaction/callbacks/test_external_callbacks.py (100%)
tests/unit_tests/external_interaction/callbacks/test_plan_reactive_callback.py (94.4%):
tests/unit_tests/external_interaction/callbacks/test_plan_reactive_callback.py:71: Cannot access member "assert_called_once" for type "function"
  Member "assert_called_once" is unknown
tests/unit_tests/external_interaction/conftest.py (100%)
-------------
Total:   400 lines
Violations: 2 lines
% Quality: 99%
-------------
```

Should we use this to help us migrate slowly to stricter type checking (and at some stage, hopefully, discard it for just using pyright)?

### PR DESCRIPTION
**Pull request by dperl-dls**
Fixes #1042 

### To test:
1. Merge this with another test branch which has errors in the diff and see the CI fail

See https://github.com/DiamondLightSource/hyperion/actions/runs/7208414314/job/19637292662

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -55,6 +55,10 @@ dev =
     tox
     build
     ruff
+    diff-cover
+    pyright
+    pyright_diff_quality_plugin @ git+https://github.com/dperl-dls/pyright_diff_quality_plugin.git
```
Comment (DominicOram) [inline setup.cfg:60]: Should: I really don't like us depending on individual's orgs like this. As in the issue can we get the code either into `diff-cover` or into the DLS org before we start depending on it?

**Review by DominicOram — CHANGES_REQUESTED** [review]
Works well, thank you! As in the issue, can you investigate putting it on the `pre-commit` too?

File: setup.cfg (hunk 1 above)
Comment (dperl-dls) [reply setup.cfg]: done. tidying it up so it has all the features to integrate into diff-cover will take a bit longer, I'd rather start using it sooner and work on that over time

**Comment by dperl-dls** [comment]
On discussion, it probably takes too long for pre-commit, and we should have pyright feedback constantly available to us in the IDE anyway

**Pull request merged**